### PR TITLE
Scoped kinds

### LINKED ISSUE #508: Use tag end field to not show as being part of previous tag
When using tags in large files with many functions there are some times when the tagbar status line will show the tag above the cursor, even if outside that tags' scope:

OK:
image

Also OK:
image

Not OK:
image

I have looked at turning on all of the tag fields for cpp and this does indeed generate an end tag, but it is ignored by tagbar:
image

The tags file was built with ```ctags --fields=* test.cpp```

If you require more information please let me know

### PR DESCRIPTION
**Pull request by raven42**
Closes #508, #516

Add --fields=e (end) to the ctags field types to look for the end of the scope.
Update the `s:GetNearbyTag()` routine to use this scope to look for the correct tag.

### CODE REVIEW

File: autoload/tagbar.vim [hunk 1]
```diff
@@ -1525,6 +1525,22 @@ function! s:ProcessTag(name, filename, pattern, fields, is_split, typeinfo, file
         let taginfo.fields.line = 0
     endif
 
+    " Make sure out 'end' is valid
```
Comment (alerque) [inline autoload/tagbar.vim:1528]:
```suggestion
    " Make sure our 'end' is valid
```

**Review by alerque — APPROVED** [review]
I looked through the code and it looks sane to me, but I haven't gotten a chance to test drive it. Use your judgement.

**Comment by raven42** [comment]
So I've noticed some performance issues with this. When moving the cursor around in a larger file in an 'unscoped' area, the cursor movement is slower now because instead of finding a tag near to the cursor, it ends up looping through all tags and doesn't find one. I'm wondering if anybody knows of a better way to try to find a match instead of looping through all tags above the current cursor location looking for a matched scope? This issue could arise even when we are inside of a scope for something like a class that has a large number of functions defined in its declaration. Then when moving the cursor around in the 'class' scope toward the bottom, but outside of any member functions it would result in this same behavior.

**Comment by raven42** [comment]
For now until any performance issues can be sorted out, I've changed the functionality so it has to be explicitly enabled by using the `'scoped-stl'` argument as a search-method for the `tagbar#currenttag()` and `tagbar#GetTagNearLine()` routines. These routines will default to the `'nearest-stl'` search method which is what they do today, but the user can optionally use the new search method to make it more intelligent at the cost of some performance.

**Review by alerque — APPROVED** [review]

**Review by alerque — APPROVED** [review]

**Pull request merged**